drivers/mfd/stm32_timers.c

```diff
@@ -0,0 +1,220 @@
+			mode = cfg->slave_trig << TIM_SMCR_TS_Pos;
+			LL_TIM_SetTriggerInput(tim, mode);
+		}
+
+		mode = cfg->master_trig << TIM_CR2_MMS_Pos;
+		LL_TIM_SetTriggerOutput(tim, mode);
+
+		/* TODO: create binding for me ?*/
+		LL_TIM_DisableARRPreload(tim);
+	}
+
+	return err;
+}
+
+static void enable(struct device *dev)
+{
+	const struct timer_stm32_config *cfg = DEV_CFG(dev);
+
+	/* Timer is enabled by master timer if slave mode is set */
+	if ((cfg->slave_mode << TIM_SMCR_SMS_Pos) != LL_TIM_SLAVEMODE_TRIGGER) {
+		struct stm32_timers *data = DEV_DATA(dev);
+		LL_TIM_EnableCounter(data->tim);
+	}
+}
+
+static void disable(struct device *dev)
+{
+	struct stm32_timers *data = DEV_DATA(dev);
+
+	LL_TIM_DisableCounter(data->tim);
+}
+
+static int get_cycles_per_sec(struct device *dev, u64_t *cycles)
+{
+	const struct timer_stm32_config *cfg = DEV_CFG(dev);
+	struct stm32_timers *data = DEV_DATA(dev);
+	u32_t bus_clk, tim_clk;
+
+	if (cycles == NULL) {
+		return -EINVAL;
+	}
+
+	/* Timer clock depends on APB prescaler */
+	clock_control_get_rate(data->clock,
+	                       (clock_control_subsys_t*)&cfg->pclken, &bus_clk);
+
+	tim_clk = __get_tim_clk(bus_clk,
+	                        (clock_control_subsys_t*)&cfg->pclken);
+
+	*cycles = (u64_t) (tim_clk / (cfg->prescaler));
+
+	return 0;
+}
+
+static const struct stm32_timers_api api = {
+	.enable = enable,
+	.disable = disable,
+	.get_cycles_per_sec = get_cycles_per_sec,
+};
+
+#define TIMER_DEVICE_INIT(n)						\
+static struct stm32_timers timer_stm32_dev_data_ ## n; 			\
+static const struct timer_stm32_config timer_stm32_dev_cfg_ ## n = {	\
+	.tim_base = DT_INST_## n ##_ST_STM32_TIMERS_BASE_ADDRESS,	\
+	.align_mode = DT_INST_## n ##_ST_STM32_TIMERS_ST_ALIGN_MODE,	\
+	.dir = DT_INST_## n ##_ST_STM32_TIMERS_ST_COUNTER_DIR_ENUM,	\
+	.msm = DT_INST_## n ##_ST_STM32_TIMERS_ST_MASTER_SLAVE_MODE,	\
+	.slave_mode = DT_INST_## n ##_ST_STM32_TIMERS_ST_SLAVE_MODE,	\
+	.slave_trig = DT_INST_## n ##_ST_STM32_TIMERS_ST_SLAVE_TRIGGER_IN, \
+	.master_trig = DT_INST_## n ##_ST_STM32_TIMERS_ST_MASTER_TRIGGER_OUT,\
+	.prescaler = DT_INST_## n ##_ST_STM32_TIMERS_ST_PRESCALER,	\
+	.pclken = { 							\
+		.bus = DT_INST_## n ##_ST_STM32_TIMERS_CLOCK_BUS,	\
+		.enr = DT_INST_## n ##_ST_STM32_TIMERS_CLOCK_BITS },	\
+};									\
+									\
+DEVICE_AND_API_INIT(timer_stm32_ ## n,					\
+		    DT_INST_## n ##_ST_STM32_TIMERS_LABEL,	 	\
+		    &init,				  		\
+		    &timer_stm32_dev_data_ ## n,			\
+		    &timer_stm32_dev_cfg_ ## n,				\
+		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,	\
+		    &api)
+
+#ifdef DT_INST_0_ST_STM32_TIMERS
+TIMER_DEVICE_INIT(0);
+#endif
+
+#ifdef DT_INST_1_ST_STM32_TIMERS
+TIMER_DEVICE_INIT(1);
+#endif
+
+#ifdef DT_INST_2_ST_STM32_TIMERS
+TIMER_DEVICE_INIT(2);
+#endif
+
+#ifdef DT_INST_3_ST_STM32_TIMERS
+TIMER_DEVICE_INIT(3);
+#endif
+
+#ifdef DT_INST_4_ST_STM32_TIMERS
+TIMER_DEVICE_INIT(4);
+#endif
+
+#ifdef DT_INST_5_ST_STM32_TIMERS
+TIMER_DEVICE_INIT(5);
+#endif
```

Comment:
A reason to limit to 5 the number of enable instances ?

Reply:
No, just had to start somewhere.